### PR TITLE
Add backup precision option

### PR DESCRIPTION
**Pull request by andrey-churkin**
### Changes

- Add a backup precision option that allows for preserving FQs on weights during the accuracy-aware algorithm.

### Reason for changes

See description in Ref: 123999

### Related tickets

Ref: 123999

### Tests

tests/openvino/native/quantization/test_quantizer_removal.py

### CODE REVIEW

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py [hunk 1]
```diff
@@ -38,6 +39,10 @@ class OVAccuracyControlAlgoBackend(AccuracyControlAlgoBackend):
 
     # Metatypes
 
+    @staticmethod
+    def get_weighted_metatypes() -> List[OVOpMetatype]:
+        return [ov_metatypes.OVMatMulMetatype, ov_metatypes.OVEmbeddingMetatype]
```
Comment (KodiaqQ) [inline nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: Why the list of the weighted metatypes is limited and not correlated with the `OPERATIONS_WITH_WEIGHTS`?

File: nncf/quantization/advanced_parameters.py [hunk 2]
```diff
@@ -130,6 +130,20 @@ class AdvancedSmoothQuantParameters:
...
+    """
+    Specifies how to revert operations to their original precision.
+
+    :param FP32: Operations will be reverted to floating-point precision.
+    :param INT8_WEIGHTS: Certain operations (such as MatMul and Embedding) will
+        be reverted to representation with int8 weights, while all other operations
+        will revert to floating-point precision.
+    """
+
+    FP32 = "fp32"
```
Comment (KodiaqQ) [inline nncf/quantization/advanced_parameters.py:143]: What about `FP16` precision?

Comment (andrey-churkin) [reply nncf/quantization/advanced_parameters.py:143]: `FP32` here means "Revert to original precision", so for `FP16` IR operations will be reverted to `FP16` precision. I agree that this name is a little bit confusing

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: I used operation metatypes to which weight compression is applied.

https://github.com/openvinotoolkit/nncf/blob/5f2c20e35cf3c1aa284275c7d4fd7ed7593a6522/nncf/quantization/algorithms/weight_compression/openvino_backend.py#L38-L39

Comment (KodiaqQ) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: Why do we focus on the `weight compression` and not the `quantize` algorithm?

File: nncf/quantization/advanced_parameters.py (hunk 2 above)
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:143]: We should rename this option then. To the `ORIGINAL` for example.

Comment (andrey-churkin) [reply nncf/quantization/advanced_parameters.py:143]: Yes, I agree. I'll rename it.

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: As far as I understand, the main idea behind the `backup_mode` parameter is to revert operations not only to their original precision but also to a representation with INT8 compressed weights. Therefore, we need to select a subset of operations with weights for which we want to preserve INT8 compressed weights. I followed the logic of the weights compression algorithm, wherein we compress weights only for `MatMul` and `Embedding` operations.

Comment (KodiaqQ) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: I don't get it.
Anyway, I'll leave the final decision to @alexsu52 and @AlexKoff88

**Comment by andrey-churkin** [comment]
I have added a description of the experiments to the ticket.

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 1 above)
Comment (alexsu52) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: Good catch! I hope we can apply this logic to all quantized operations not only for operation which support for weights compression, if it does not lead to performance degradation.

File: nncf/quantization/advanced_parameters.py [hunk 3]
```diff
@@ -222,13 +236,16 @@ class AdvancedAccuracyRestorerParameters:
...
+    :param backup_mode: Specifies how to revert operations to their original precision.
+    :type backup_mode: BackupMode
     """
 
     max_num_iterations: int = sys.maxsize
     tune_hyperparams: bool = False
     ranking_subset_size: Optional[int] = None
     num_ranking_workers: Optional[int] = None
     intermediate_model_dir: Optional[str] = None
+    backup_mode: BackupMode = BackupMode.FP32
```
Comment (alexsu52) [inline nncf/quantization/advanced_parameters.py:248]: @andrey-churkin @KodiaqQ What do you think about this proposal?
```
mode: RestoreMode = RestoreMode.ACTIVATIONS_AND_WEIGHTS


class RestoreMode(Enum):
    ACTIVATIONS_AND_WEIGHTS = "activations_and_weights"
    ONLY_ACTIVATIONS = "only_activations"
```
Using QuantizerGroup is more flexible and is not introduced new class but give user a chance to remove FQ from weights:

```
groups: List[QuantizerGroup] = [QuantizerGroup.ACTIVATIONS, QuantizerGroup.WEIGHTS]
```


cc` @AlexKoff88

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: nncf/quantization/advanced_parameters.py (hunk 3 above)
Comment (KodiaqQ) [reply nncf/quantization/advanced_parameters.py:248]: LGTM.

File: nncf/quantization/advanced_parameters.py (hunk 2 above)
Comment (andrey-churkin) [reply nncf/quantization/advanced_parameters.py]: Done

File: nncf/quantization/advanced_parameters.py (hunk 3 above)
Comment (andrey-churkin) [reply nncf/quantization/advanced_parameters.py]: Done

File: nncf/quantization/algorithms/accuracy_control/openvino_backend.py (hunk 1 above)
Comment (andrey-churkin) [reply nncf/quantization/algorithms/accuracy_control/openvino_backend.py:44]: I reverted half of all convolutions in MobileNetV2 using both `RestoreMode.ACTIVATIONS_AND_WEIGHTS` and `RestoreMode.ONLY_ACTIVATIONS`. There was no difference in performance between the resulting models, so I believe we can specify `OPERATIONS_WITH_WEIGHTS` here.

**Review by KodiaqQ — APPROVED** [review]

File: nncf/quantization/advanced_parameters.py [hunk 4]
```diff
@@ -130,6 +130,20 @@ class AdvancedSmoothQuantParameters:
     matmul: float = 0.95
 
 
+class RestoreMode(Enum):
+    """
+    Specifies how to revert operations to their original precision.
+
+    :param ACTIVATIONS_AND_WEIGHTS: Operations will be reverted to floating-point precision.
+    :param ONLY_ACTIVATIONS: Certain operations (such as MatMul and Embedding) will
```
Comment (alexsu52) [inline nncf/quantization/advanced_parameters.py:138]: It looks like you need update the docstring taking into account https://github.com/openvinotoolkit/nncf/pull/2299#discussion_r1432667080

Comment (andrey-churkin) [reply nncf/quantization/advanced_parameters.py]: Thanks! Fixed.

**Review by alexsu52 — APPROVED** [review]
LGTM

**Pull request merged**